force_gate_controller/src/force_gate_parent.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020 PAL Robotics S.L.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// #include <string>
+
+#include "controller_interface/controller_interface.hpp"
+// #include "hardware_interface/types/hardware_interface_type_values.hpp"
+#include "force_gate_controller/force_gate_parent.hpp"
+#include "rclcpp/logging.hpp"
+// #include "rclcpp/parameter.hpp"
+
+namespace force_gate_controller
+{
+
+controller_interface::CallbackReturn ForceGateParent::read_force_gate_parameters(std::shared_ptr<rclcpp_lifecycle::LifecycleNode> node)
+{
+  if (!force_gate_param_listener_)
+  {
+    force_gate_param_listener_ = std::make_shared<ParamListener>(node);
+    if (!force_gate_param_listener_) {
+      RCLCPP_ERROR(node->get_logger(), "Error encountered during init");
+      return controller_interface::CallbackReturn::ERROR;
+    }
+  }
+  force_gate_params_ = force_gate_param_listener_->get_params();
+
+  // Update wrench tolerances if thresholding enabled
+  rt_wrench_stamped_.writeFromNonRT(nullptr);
+  if (force_gate_params_.wrench_threshold.topic != "")
+  {
+    RCLCPP_INFO(node->get_logger(), "Updating Wrench Thresholds");
+
+    wrench_tolerances_.timeout = rclcpp::Duration::from_seconds(force_gate_params_.wrench_threshold.timeout);
+    wrench_tolerances_.forceTotal = force_gate_params_.wrench_threshold.fMag;
+    wrench_tolerances_.forceVec[0] = force_gate_params_.wrench_threshold.fx;
+    wrench_tolerances_.forceVec[1] = force_gate_params_.wrench_threshold.fy;
+    wrench_tolerances_.forceVec[2] = force_gate_params_.wrench_threshold.fz;
+    wrench_tolerances_.torqueTotal = force_gate_params_.wrench_threshold.tMag;
+    wrench_tolerances_.torqueVec[0] = force_gate_params_.wrench_threshold.tx;
+    wrench_tolerances_.torqueVec[1] = force_gate_params_.wrench_threshold.ty;
+    wrench_tolerances_.torqueVec[2] = force_gate_params_.wrench_threshold.tz;
+
+    // Subscribe to wrench topic
+    wrench_subscriber_ = node->create_subscription<geometry_msgs::msg::WrenchStamped>(
```

Comment:
Force gate JTC makes the subscriber once in `on_configure()`. This will re-make it every time the parameters are read/updated. I would wrap this in an `if wrench_subscriber_ != nullptr`.

Recall that the topic is defined to be read-only in the parameters file, so you should never have to re-make the subscriber in the lifetime of the controller.